Corvus.Extensions.Samples/TraversalExtensionsSample.dib

```diff
@@ -0,0 +1,242 @@
+#!markdown
+
+## Traversal Extensions - Sample Notebook
+
+#!markdown
+
+First, add a reference to the `Corvus.Extensions` NuGet package.
+
+#!csharp
+
+#r "nuget: Corvus.Extensions, 1.1.4"
+
+using Corvus.Extensions;
+
+#!markdown
+
+### ForEachAsync()
+
+#!markdown
+
+Execute an async action for each item in the enumerable.
+
+#!csharp
+
+public async Task IsEqualToOne(int x)
+{
+    bool result = x.Equals(1);
+    Console.WriteLine(result);
+}
+
+#!csharp
+
+var list = new List<int> { 1, 2, 3, 4, 5 };
+await list.ForEachAsync<int>(IsEqualToOne)
+
+#!markdown
+
+The `ForEachAsync()` method is here for historical interest. It is now possible to write the following and get the same output:
+
+#!csharp
+
+foreach (int x in list)
+{
+    await IsEqualToOne(x);
+}
+
+#!markdown
+
+### ForEachAtIndex<T>()
+
+#!markdown
+
+Execute an action for each item in the enumerable with the index of the item in the enumerable.
+
+#!csharp
+
+static void ConsolePrint(string s, int i)
+{
+    Console.WriteLine(i);
+    Console.WriteLine(s);
+}
+
+Action<string, int> printAction = ConsolePrint;
+
+#!csharp
+
+IEnumerable<string> list = new List<string> { "foo", "bar", "quux" };
+list.ForEachAtIndex<string>(printAction);
+
+#!csharp
+
+int i = 0;
+
+foreach (string v in list)
+{
+    printAction(v, i++);
+}
+
+#!markdown
+
+### ForEachAtIndexAsync()
+
+#!markdown
+
+Execute an async action for each item in the enumerable, in turn, with the index of the item in the enumerable.
+
+#!csharp
+
+static async Task ConsolePrint(string s, int i)
+{
+    Console.WriteLine(s);
+    Console.WriteLine(i);
+}
+
+#!csharp
+
+IEnumerable<string> list = new List<string> { "foo", "bar", "quux" };
+await list.ForEachAtIndexAsync<string>(ConsolePrint);
+
+#!markdown
+
+### ForEachFailEnd()
+
+#!markdown
+
+Execute an action for each item in the enumerable. 
+
+If any operation fails, then the enumeration is continued to the end when an Aggregate Exception is thrown containing the exceptions thrown by any failed operations.
+
+#!csharp
+
+IEnumerable<int> list = new List<int> { 1, 2, 8, 0, 5 };
+
+static void Divide(int x)
+{
+    int result = 10/x;
+    Console.WriteLine(result);
+}
+
+Action<int> divideAction = Divide;
+
+list.ForEachFailEnd<int>(divideAction)
+
+#!markdown
+
+### ForEachFailEndAsync()
+
+#!markdown
+
+Execute an async action for each item in the enumerable.
+
+Returns a task which completes when the enumeration has completed.
+
+If any operation fails, then the enumeration is continued to the end when an Aggregate Exception is thrown containing the exceptions thrown by any failed operations.
+
+#!csharp
+
+IEnumerable<int> list = new List<int> { 1, 2, 8, 0, 5 };
+
+public async Task DivideAsync(int x)
+{
+    int result = 10/x;
+    Console.WriteLine(result);
+}
```

Comment:
Again, bear in mind that this code would cause VS to give you a warning. When code is given as an example of how to use something, it's not going to add confidence when people find the compiler doesn't like it.